pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1641,23 +1643,30 @@ public void addNewSegment(String tableName, SegmentMetadata segmentMetadata, Str
       Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap = Collections
           .singletonMap(InstancePartitionsType.OFFLINE, InstancePartitionsUtils
               .fetchOrComputeInstancePartitions(_helixZkManager, offlineTableConfig, InstancePartitionsType.OFFLINE));
-      HelixHelper.updateIdealState(_helixZkManager, offlineTableName, idealState -> {
-        assert idealState != null;
-        Map<String, Map<String, String>> currentAssignment = idealState.getRecord().getMapFields();
-        if (currentAssignment.containsKey(segmentName)) {
-          LOGGER.warn("Segment: {} already exists in the IdealState for table: {}, do not update", segmentName,
-              offlineTableName);
-        } else {
-          List<String> assignedInstances =
-              segmentAssignment.assignSegment(segmentName, currentAssignment, instancePartitionsMap);
-          LOGGER.info("Assigning segment: {} to instances: {} for table: {}", segmentName, assignedInstances,
-              offlineTableName);
-          currentAssignment.put(segmentName,
-              SegmentAssignmentUtils.getInstanceStateMap(assignedInstances, SegmentStateModel.ONLINE));
+      if (_tableUpdaterLockMap.get(offlineTableName) == null) {
+        synchronized (_tableUpdaterLockMap) {
+          if (_tableUpdaterLockMap.get(offlineTableName) == null) {
+            _tableUpdaterLockMap.put(offlineTableName, new Object());
```

Comment:
This will increase the number of tables in the map forever. Can you do a similar approach as is done in PinotLLCRealtimeSegmentManager? Maybe you can use the same config, or a hard-coded number of locks for now.

Reply:
It will be one per table, I feel it should be ok. The purpose of adding  this lock is to ensure we only update one segment per table at a time and we can still support multiple tables to be updated simultaneously.

Reply:
Is this for offline segments? If so, how can this prevent other controllers from updating the idealstate? 

For realtime, a lock makes sense since there is one leader for each table that handles idealstate update. Other controllers will handle different tables. This is not the case with offline segments.

Reply:
This won't prevent from other controller, just to improve on the single controller update. In practice, it will help a lot on idealstate update as well, as we have only 3 parallel writes to update idealstates instead of segment push job parallelism.

From client side, we can also try to push to one single controller to avoid the race condition.

Reply:
Also, we hold the lock while backing off, so a single cross-controller contention will start building a queue? I think a better solution may to simply reduce the backoff time and try more often?

Also, we have clusters with 1000s of tables, and in this case, the hash table will contain all of them.

Reply:
@mcvsubbu A map with less than 10000 entries should be tiny (less than 1MB). Actually the `_segmentCrcMap` and `_lastKnownSegmentMetadataVersionMap` might cause problem because it stores an entry per segment, but that is out of the scope of this PR

Reply:
Discussing offline, I gather this:

- This helps in case of thousands of segments being pushed. For metadata push, we still need to generate the metadata from segment tar file, which takes time. In order to improve that, if we increase parallelism, then metadata generation improves. However, parallel push in segments start taking longer due to retries. If we reduce parallelism, the metadata generation slows down as well.

- With this PR, we can still keep high parallelism for metadata generation, but also reduce longer update time (or even failures), by reducing the probability of retries.

- The Map size should be small, and the synchronization is only for IS update. So this can be thought of as a bug fix, which is reducing number of retries, without penalizing on push time with parallelism.

We also discussed the following:
- We can also reduce number of IS updates (and hence retries) by having bulk updates to IS. For example, metadata push can pack multiple segment metadatas in one POST call.